src/integrations/third-party/the-events-calendar/EventsCalendarSchema.php

```diff
@@ -0,0 +1,241 @@
+
+		if ( is_singular( 'tribe_events' ) ) {
+			global $post;
+			$posts[] = $post;
+		} elseif (
+			tribe_is_month()
+		) {
+			$posts = $this->get_month_events();
+		}
+
+		$tribe_data = $this->get_tribe_schema( $posts );
+		$tribe_data = $this->transform_tribe_schema( $tribe_data );
+
+		$data = array();
+		foreach ( $tribe_data as $t ) {
+			// Cast the schema object as array, the Yoast Class can't handle objects.
+			$data[] = (array) $t;
+		}
+
+		// If the resulting array only has one entry, print it directly.
+		if ( count( $data ) === 1 ) {
+			$data                     = $data[0];
+			$data['mainEntityOfPage'] = [ '@id' => $this->context->canonical . Schema_IDs::WEBPAGE_HASH ];
+		} elseif ( count( $data ) === 0 ) {
+			$data = false;
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Get and return the schema markup for a collection of posts.
+	 * If the posts array is empty, only the current post is returned.
+	 *
+	 * @param  array $posts The collection of posts we want schema markup for.
+	 *
+	 * @return array        The tribe schema for these posts.
+	 */
+	private function get_tribe_schema( array $posts = [] ) {
+		$args       = array(
+			// We do not want the @context to be shown.
+			'context' => false,
+		);
+		$tribe_data = Tribe__Events__JSON_LD__Event::instance()->get_data( $posts, $args );
+		$type       = strtolower( esc_attr( Tribe__Events__JSON_LD__Event::instance()->type ) );
+
+		foreach ( $tribe_data as $post_id => $_data ) {
+			Tribe__Events__JSON_LD__Event::instance()->set_type( $post_id, $type );
+			// Register this post as done already.
+			Tribe__Events__JSON_LD__Event::instance()->register( $post_id );
+		}
+
+		/**
+		 * Allows the event data to be modifed by themes and other plugins.
+		 *
+		 * @example yoast_tec_json_ld_thing_data
+		 * @example yoast_tec_json_ld_event_data
+		 *
+		 * @param array $data objects representing the Google Markup for each event.
+		 * @param array $args the arguments used to get data
+		 */
+		$tribe_data = apply_filters( "yoast_tec_json_ld_{$type}_data", $tribe_data, $args );
+
+		return $tribe_data;
+	}
+
+	/**
+	 * Transform the tribe schema markup and adapt it to the Yoast SEO standard.
+	 *
+	 * @param  array $data The data retrieved from the TEC plugin.
+	 *
+	 * @return array       The transformed event data.
+	 */
+	private function transform_tribe_schema( array $data = [] ) {
+		$new_data = array();
+
+		foreach ( $data as $post_id => $d ) {
+			/*
+			 * EVENT
+			 */
+			// Generate an @id for the event.
+			$d->{'@id'} = get_permalink( $post_id ) . '#' . strtolower( esc_attr( $d->{'@type'} ) );
+
+			// Transform the post_thumbnail from the url to the @id of #primaryimage.
+			if ( has_post_thumbnail( $post_id ) ) {
+				if ( is_singular( 'tribe_events' ) ) {
+					// On a single view we can assume that Yoast SEO already printed the
+					// image schema for the post thumbnail.
+					$d->image = (object) [
+						'@id' => get_permalink( $post_id ) . '#primaryimage',
```

Comment:
The permalink for this post_id was already calculated above. I would strongly suggest storing it in a variable to be able to reuse it here to avoid duplicate calculations.